### PR TITLE
Standardize utils and remove old proxies.

### PR DESCRIPTION
**Pull request by paulfalgout**
This is a lot.. not sure how to break it up yet, but this PR is overkill.. and there's some tests failing I need to check into..

But essentially this was looking at how all of the various utils and mixed in functions could be made more similar.

There's a couple of rabbit trails related to similar consistency concerns.. but I just wanted to open a PR to see if this was a reasonable direction before I fix it.

### CODE REVIEW

**Comment by paulfalgout** [comment]
Ok.  I think this at a good state.

It introduces 2 new utils for internal use 
- `_setOptions` which is for the constructors and merges and attaches options to the object.. in a couple of cases we were just `this.options = options` so that we could use a `this.getOption` prior to a `Marionette.Object.call`, but that actually breaks the ability to predefine options on the definition of those objects.
- `proxy` which essentially relieves the need for the various proxy functions but still allows for the generic util attached to `Marionette` whose first argument is the target.

Additionally I've made `getValue` public and have changed the API so that it works like the others and expects to be added to a prototype or `proxy`'d on to `Marionette`.  Also it passes any addition ...args to the method passed to it, where-as it used to take an array of the arguments to pass to the method.

It also adds a CommonMixin for the utils common to every Mn class.

And it simplifies `triggerMethod` in ways doomed to conflict with https://github.com/marionettejs/backbone.marionette/pull/2845

@marionettejs/marionette-core :eyes:   This one probably shouldn't hang around too long.. Lemme know if adjustments need to be made.

**Comment by paulfalgout** [comment]
This'll resolve https://github.com/marionettejs/backbone.marionette/issues/2814

**Comment by jasonLaster** [comment]
Great deep dive. Utils look 10x better. Every class is so much easier on the eyes. 

:thumbsup:

I'll merge today if there are no objections. This is big, but there are only a couple key ideas applied at scale.

**Comment by paulfalgout** [comment]
@jasonLaster FWIW it might be more interesting now to reopen the idea of `getOptionValue` some of the  consolidating of styles made it such that there are more places where it'd be useful.. 
Even `triggerMethod` could become:
```js
function triggerMethod(event, ...args) {
  // get the method name from the event name
  var methodName = 'on' + event.replace(splitter, getEventName);
  var result = getOptionValue.call(this, methodName, ...args);
  
  // trigger the event
  this.trigger(event, ...args);

  return result;
}
```
That _might_ change the return value of `triggerMethod`, in some strange cases if you were to have an `onEvent` defined that isn't a function, but otherwise functionally the same.

Also I haven't analyzed the output at all as far as what babel transpiles to.. how concerned, or with what functions should we be concerned about the output here?

**Comment by paulfalgout** [comment]
Checking into how some of the spreads transpile.. and hmmm :disappointed:   

https://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&code=var%20method%20%3D%20function()%7B%7D%3B%0Afunction%20triggerMethod(...args)%7B%0Amethod.call(this%2C%20...args)%3B%0A%7D

**Comment by megawac** [comment]
Yeah, if we're bumping the version range of _ to 1.8/3.* we could also use
restParam

**Comment by jasonLaster** [comment]
@paulfalgout, this is a great win for the readability of all of our common classes. Also, it'll make it much easier to do `getOptionValue` or other small utils.

Not so worried about the babel splat code. The way I see it, if we cared enough we could probably override that function with an `_.toArray(arguments)` call `babel-plugin-arguments-to-array-ftw`

**Comment by megawac** [comment]
@jasonlaster if we decide we care we should just replace splats with
https://lodash.com/docs#restParam

File: src/mixins/behaviors.js [hunk 1]
```diff
@@ -87,12 +86,11 @@ export default {
...
   },
 
-  _triggerEventOnBehaviors: function(args) {
-    var triggerMethod = _triggerMethod;
+  _triggerEventOnBehaviors: function(...args) {
     var behaviors = this._behaviors;
     // Use good ol' for as this is a very hot function
     for (var i = 0, length = behaviors && behaviors.length; i < length; i++) {
-      triggerMethod(behaviors[i], args);
+      triggerMethod.call(behaviors[i], ...args);
```
Comment (megawac) [inline src/mixins/behaviors.js]: This seems very, very weird. Should this not be `triggerMethod.apply(behaviours[i], args)`

File: src/mixins/view.js [hunk 2]
```diff
@@ -236,38 +228,37 @@ export default {
 
   // import the `triggerMethod` to trigger events with corresponding
   // methods if the method exists
-  triggerMethod: function(eventName, ...args) {
-    var ret = _triggerMethod(this, arguments);
+  triggerMethod: function(...args) {
+    var ret = triggerMethod.call(this, ...args);
```
Comment (megawac) [inline src/mixins/view.js]: Same here, you want `triggerMethod.apply(this, arguments)`

**Comment by megawac** [comment]
Looks good, just remove uses of applied splats with `.call`

**Pull request merged**